recon/src/libp2p.rs

```diff
@@ -0,0 +1,232 @@
+            libp2p::swarm::FromSwarm::ListenerError(_) => {}
+            libp2p::swarm::FromSwarm::ListenerClosed(_) => {}
+            libp2p::swarm::FromSwarm::NewExternalAddr(_) => {}
+            libp2p::swarm::FromSwarm::ExpiredExternalAddr(_) => {}
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: iroh_p2p::PeerId,
+        _connection_id: libp2p::swarm::ConnectionId,
+        event: libp2p::swarm::THandlerOutEvent<Self>,
+    ) {
+        match event {
+            // The peer has synchronized with us, mark the time and record that the peer conneciton
+            // is now idle.
+            FromHandler::PeerSynchronized => self.peers.entry(peer_id).and_modify(|info| {
+                info.last_sync = Some(Instant::now());
+                info.status = PeerStatus::Idle;
+            }),
+        };
+    }
+
+    fn poll(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+        _params: &mut impl libp2p::swarm::PollParameters,
+    ) -> std::task::Poll<libp2p::swarm::ToSwarm<Self::OutEvent, libp2p::swarm::THandlerInEvent<Self>>>
+    {
+        // Check each peer and start synchronization as needed.
+        for (peer_id, info) in &mut self.peers {
+            match info.status {
+                PeerStatus::Idle => {
+                    let should_sync = if let Some(last_sync) = &info.last_sync {
+                        last_sync.elapsed() > Duration::from_millis(1000)
+                    } else {
+                        true
+                    };
+                    if should_sync {
+                        info.status = PeerStatus::Started;
+                        return Poll::Ready(ToSwarm::NotifyHandler {
+                            peer_id: *peer_id,
+                            handler: NotifyHandler::One(info.connection_id),
+                            event: FromBehaviour::StartSync,
+                        });
+                    }
+                }
+                PeerStatus::Started => {}
+            }
+        }
+        // TODO remove this demo code
+        // Randomly insert a new key into the Recon key space.
+        if rand::thread_rng().gen::<f64>() > 0.9 {
+            debug!("adding new key");
+            let s: String = rand::thread_rng()
+                .sample_iter(&Alphanumeric)
+                .take(12)
+                .map(char::from)
+                .collect();
+            self.recon.insert_key(&s);
+        }
+        Poll::Pending
+    }
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: libp2p::swarm::ConnectionId,
+        peer: PeerId,
+        _local_addr: &libp2p::Multiaddr,
+        _remote_addr: &libp2p::Multiaddr,
+    ) -> std::result::Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
+        debug!(%peer, "handle_established_inbound_connection");
+        Ok(Handler::new(self.recon.clone()))
```

Comment:
Is this cloning the whole reason b-tree  or just the pointer?

Reply:
just the pointer